core/net/ip/tcpip.c

```diff
@@ -276,44 +288,30 @@ tcp_listen(uint16_t port)
 }
 /*---------------------------------------------------------------------------*/
 void
-tcp_attach(struct uip_conn *conn,
-	   void *appstate)
+tcp_attach(struct uip_conn *conn, void *appstate)
 {
-  uip_tcp_appstate_t *s;
-
-  s = &conn->appstate;
-  s->p = PROCESS_CURRENT();
-  s->state = appstate;
+  setup_appstate(&conn->appstate, appstate);
 }
-
 #endif /* UIP_TCP */
 /*---------------------------------------------------------------------------*/
 #if UIP_UDP
 void
-udp_attach(struct uip_udp_conn *conn,
-	   void *appstate)
+udp_attach(struct uip_udp_conn *conn, void *appstate)
 {
-  uip_udp_appstate_t *s;
-
-  s = &conn->appstate;
-  s->p = PROCESS_CURRENT();
-  s->state = appstate;
+  setup_appstate(&conn->appstate, appstate);
 }
 /*---------------------------------------------------------------------------*/
 struct uip_udp_conn *
 udp_new(const uip_ipaddr_t *ripaddr, uint16_t port, void *appstate)
 {
   struct uip_udp_conn *c;
-  uip_udp_appstate_t *s;
 
   c = uip_udp_new(ripaddr, port);
```

Comment:
The cc65 compiler has that problem, but only with structs and unions.  It can handle non-constant expressions in the  initiations of "scalar" variables.